sympy/matrices/expressions/tests/test_matexpr.py

```diff
@@ -342,7 +341,7 @@ def test_inv():
     B = MatrixSymbol('B', 3, 3)
     assert B.inv() == B**-1
 
-
+@XFAIL
```

Comment:
you are adding some `XFAIL` s, are you sure it's a good idea?

Reply:
There are two XFAILs added here. I have mentioned the reasoning in the commit messages, but I will reiterate here. 

This XFAIL has to do with functions like `factor` no longer working with matrix expressions. In fact, many simplification functions now raise `NotImplementedError("noncommutative scalars in MatMul are not supported.")` (in some instances, you have to merge https://github.com/sympy/sympy/pull/15121 to see this message). This message results from algorithms that try to substitute matrix symbols with noncommutative symbols. This is fundamentally broken, because MatMul separates noncommutative symbols as the coefficient, effectively reordering them. You can see the commit message of https://github.com/sympy/sympy/pull/15948/commits/b98319aa86bdc2917e6fe97906944e2556478a43 an example where `factor` was returning a wrong result. I decided it was better for these functions to raise NotImplementedError rather than give wrong results in some cases. Basically, each algorithm will need to be written explicitly with matrix expressions in mind, or else some higher level fix would be needed (maybe make MatMul.subs(Matrix, noncommutative Symbol) do something smarter). Most of these already don't work in master because of https://github.com/sympy/sympy/pull/15121 (see https://github.com/sympy/sympy/issues/15120 and https://github.com/sympy/sympy/issues/6249), so I don't see an issue with XFAILing these for now. 

The other XFAIL is below. It is that Add(nan, matrix) returns nan. This is because of the way Add does its nan short circuiting. It didn't deem it worthy to make this work, since it would require refactoring Add.flatten (right now it unconditionally returns nan whenever it sees a nan). If it ever becomes an issue, we can go in and fix it.

Reply:
Oh and I forgot the third XFAIL in the postprocessor tests. This is because I removed the ability for postprocessors to work on subexpressions. This is necessary for performance purposes. The current implementation also produces a memory leak. Since the API is not used anywhere else in library code, and it is explicitly designated as experimental, I think we can make such changes to it. 

Actually I have another suggestion for an even more dramatic change to the API [here](https://github.com/asmeurer/sympy/commit/b0777d925d005bf5973702033a7147bda09a1b64). I haven't pushed it to this branch yet because I want to know if it is actually more performant or not. If it isn't, it likely isn't worth the change (even though the API is itself much cleaner IMO).